// prosemirror/dist/serialize/markdown.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

exports.toMarkdown = toMarkdown;

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var _model = require("../model");

var _index = require("./index");

function toMarkdown(doc) {
  var state = new State();
  state.renderContent(doc);
  return state.out;
}

(0, _index.defineTarget)("markdown", toMarkdown);

function esc(str, startOfLine) {
  str = str.replace(/[`*\\~+\[\]]/g, "\\$&");
  if (startOfLine) str = str.replace(/^[:#-]/, "\\$&");
  return str;
}

function quote(str) {
  var wrap = str.indexOf('"') == -1 ? '""' : str.indexOf("'") == -1 ? "''" : "()";
  return wrap[0] + str + wrap[1];
}

function rep(str, n) {
  var out = "";
  for (var i = 0; i < n; i++) {
    out += str;
  }return out;
}

var State = (function () {
  function State() {
    _classCallCheck(this, State);

    this.delim = this.out = "";
    this.closed = false;
    this.inTightList = false;
  }

  _createClass(State, [{
    key: "closeBlock",
    value: function closeBlock(node) {
      this.closed = node;
    }
  }, {
    key: "flushClose",
    value: function flushClose(size) {
      if (this.closed) {
        if (!this.atBlank()) this.out += "\n";
        if (size == null) size = 2;
        if (size > 1) {
          var delimMin = this.delim;
          var trim = /\s+$/.exec(delimMin);
          if (trim) delimMin = delimMin.slice(0, delimMin.length - trim[0].length);
          for (var i = 1; i < size; i++) {
            this.out += delimMin + "\n";
          }
        }
        this.closed = false;
      }
    }
  }, {
    key: "wrapBlock",
    value: function wrapBlock(delim, firstDelim, node, f) {
      var old = this.delim;
      this.write(firstDelim || delim);
      this.delim += delim;
      f();
      this.delim = old;
      this.closeBlock(node);
    }
  }, {
    key: "atBlank",
    value: function atBlank() {
      return (/(^|\n)$/.test(this.out)
      );
    }
  }, {
    key: "write",
    value: function write(add) {
      this.flushClose();
      if (this.delim && this.atBlank()) this.out += this.delim;
      if (add) this.out += add;
    }
  }, {
    key: "text",
    value: function text(_text, escape) {
      var lines = _text.split("\n");
      for (var i = 0; i < lines.length; i++) {
        var startOfLine = this.atBlank() || this.closed;
        this.write();
        this.out += escape !== false ? esc(lines[i], startOfLine) : lines[i];
        if (i != lines.length - 1) this.out += "\n";
      }
    }
  }, {
    key: "ensureNewLine",
    value: function ensureNewLine() {
      if (!this.atBlank()) this.out += "\n";
    }
  }, {
    key: "render",
    value: function render(node) {
      node.type.serializeMarkdown(this, node);
    }
  }, {
    key: "renderContent",
    value: function renderContent(node) {
      for (var i = 0; i < node.size; i++) {
        this.render(node.child(i));
      }
    }
  }, {
    key: "renderInline",
    value: function renderInline(parent) {
      var _this = this;

      var stack = [];
      var progress = function progress(node) {
        var marks = node ? node.marks.slice() : [];
        if (stack.length && stack[stack.length - 1].type == "code" && (!marks.length || marks[marks.length - 1].type != "code")) {
          _this.text("`", false);
          stack.pop();
        }
        for (var j = 0; j < stack.length; j++) {
          var cur = stack[j],
              found = false;
          for (var k = 0; k < marks.length; k++) {
            if (marks[k].eq(stack[j])) {
              marks.splice(k, 1);
              found = true;
              break;
            }
          }
          if (!found) {
            _this.text(markString(cur, false), false);
            stack.splice(j--, 1);
          }
        }
        for (var j = 0; j < marks.length; j++) {
          var cur = marks[j];
          stack.push(cur);
          _this.text(markString(cur, true), false);
        }
        if (node) _this.render(node);
      };
      parent.forEach(progress);
      progress(null);
    }
  }, {
    key: "renderList",
    value: function renderList(node, delim, firstDelim) {
      var _this2 = this;

      if (this.closed && this.closed.type == node.type) this.flushClose(3);else if (this.inTightList) this.flushClose(1);

      var prevTight = this.inTightList;
      this.inTightList = node.attrs.tight;

      var _loop = function (i, n, item) {
        if (n && node.attrs.tight) _this2.flushClose(1);
        _this2.wrapBlock(delim, firstDelim(n), node, function () {
          return _this2.render(item);
        });
      };

      for (var i = node.iter(), n = 0, item = undefined; item = i.next().value; n++) {
        _loop(i, n, item);
      }
      this.inTightList = prevTight;
    }
  }]);

  return State;
})();

function def(cls, method) {
  cls.prototype.serializeMarkdown = method;
}

def(_model.BlockQuote, function (state, node) {
  state.wrapBlock("> ", null, node, function () {
    return state.renderContent(node);
  });
});

def(_model.CodeBlock, function (state, node) {
  if (node.attrs.params == null) {
    state.wrapBlock("    ", null, node, function () {
      return state.text(node.textContent, false);
    });
  } else {
    state.write("```" + node.attrs.params + "\n");
    state.text(node.textContent, false);
    state.ensureNewLine();
    state.write("```");
    state.closeBlock(node);
  }
});

def(_model.Heading, function (state, node) {
  state.write(rep("#", node.attrs.level) + " ");
  state.renderInline(node);
  state.closeBlock(node);
});

def(_model.HorizontalRule, function (state, node) {
  state.write(node.attrs.markup || "---");
  state.closeBlock(node);
});

def(_model.BulletList, function (state, node) {
  state.renderList(node, "  ", function () {
    return (node.attrs.bullet || "*") + " ";
  });
});

def(_model.OrderedList, function (state, node) {
  var start = Number(node.attrs.order || 1);
  var maxW = String(start + node.size - 1).length;
  var space = rep(" ", maxW + 2);
  state.renderList(node, space, function (i) {
    var nStr = String(start + i);
    return rep(" ", maxW - nStr.length) + nStr + ". ";
  });
});

def(_model.ListItem, function (state, node) {
  return state.renderContent(node);
});

def(_model.Paragraph, function (state, node) {
  state.renderInline(node);
  state.closeBlock(node);
});

// Inline nodes

def(_model.Image, function (state, node) {
  state.write("![" + esc(node.attrs.alt || "") + "](" + esc(node.attrs.src) + (node.attrs.title ? " " + quote(node.attrs.title) : "") + ")");
});

def(_model.HardBreak, function (state) {
  return state.write("\\\n");
});

def(_model.Text, function (state, node) {
  return state.text(node.text);
});

// Marks

function markString(mark, open) {
  var value = open ? mark.type.openMarkdown : mark.type.closeMarkdown;
  return typeof value == "string" ? value : value(mark);
}

function defMark(mark, open, close) {
  mark.prototype.openMarkdown = open;
  mark.prototype.closeMarkdown = close;
}

defMark(_model.EmMark, "*", "*");

defMark(_model.StrongMark, "**", "**");

defMark(_model.LinkMark, "[", function (mark) {
  return "](" + esc(mark.attrs.href) + (mark.attrs.title ? " " + quote(mark.attrs.title) : "") + ")";
});

defMark(_model.CodeMark, "`", "`");
